### PR TITLE
Introduce AWS provider s3 directory path variables

### PR DESCRIPTION
**Pull request by remi00**
**What it does?**

- Introduce AWS provider s3 directory path variables that will allow for proper directory structure within artifacts S3 bucket once hash artifacts versioning will be added

**What else is important?**

- Implements [this suggestion](https://github.com/serverless/serverless/pull/9876#discussion_r698418058) from PR #9876

### CODE REVIEW

**Comment by remi00** [comment]
Hello @medikoo could you please take a look? Do these changes make sense from your point of view?

File: lib/plugins/aws/provider.js [hunk 1]
```diff
@@ -1592,6 +1593,26 @@ class AwsProvider {
...
   }
 
+  initS3DirectoryPaths(provider) {
+    const date = new Date();
+    const serviceStage = `${this.serverless.service.service}/${this.getStage()}`;
+    const dateString = `${date.getTime().toString()}-${date.toISOString()}`;
+    const prefix = this.provider.getDeploymentPrefix();
+
+    provider.s3CodeArtifactsDirectoryPath = `${prefix}/${serviceStage}/code-artifacts`;
+    provider.s3DeploymentDirectoryPath = `${prefix}/${serviceStage}/${dateString}`;
```
Comment (medikoo) [inline lib/plugins/aws/provider.js:1603]: Let's not set it on `service.provider`, this object reflects user configuration and ideally if we do not mix it with internally resolved data (although I know that currently our internals in few places do that).

Ideally if `s3CodeArtifactsDirectoryPath` and `s3DeploymentDirectoryPath` getters simply resolve it on spot
And _timestamp_ part can be set on provider instance in its constructor as e.g. `this.deploymentTimestamp = ...`

File: lib/plugins/aws/provider.js [hunk 2]
```diff
@@ -1592,6 +1593,26 @@ class AwsProvider {
     return `${provider.deploymentPrefix}`;
   }
 
+  initS3DirectoryPaths(provider) {
+    const date = new Date();
+    const serviceStage = `${this.serverless.service.service}/${this.getStage()}`;
+    const dateString = `${date.getTime().toString()}-${date.toISOString()}`;
+    const prefix = this.provider.getDeploymentPrefix();
+
+    provider.s3CodeArtifactsDirectoryPath = `${prefix}/${serviceStage}/code-artifacts`;
```
Comment (medikoo) [inline lib/plugins/aws/provider.js:1602]: In context of this PR we should not yet introduce `s3CodeArtifactsDirectoryPath`

File: lib/plugins/aws/provider.js [hunk 3]
```diff
@@ -1592,6 +1593,26 @@ class AwsProvider {
...
+    provider.s3DeploymentDirectoryPath = `${prefix}/${serviceStage}/${dateString}`;
+  }
+
+  get s3CodeArtifactsDirectoryPath() {
```
Comment (medikoo) [inline lib/plugins/aws/provider.js:1606]: In context of this PR we should not yet introduce `s3CodeArtifactsDirectoryPath`, as so far path is same as `s3DeploymentDirectoryPath`.

File: lib/plugins/aws/rollback.js [hunk 4]
```diff
@@ -77,7 +77,7 @@ class AwsRollback {
       throw new ServerlessError(`${msg} ${hint}`, 'ROLLBACK_DEPLOYMENT_NOT_FOUND');
     }
 
-    service.package.artifactDirectoryName = `${prefix}/${dateString}`;
+    service.provider.s3DeploymentDirectoryPath = `${prefix}/${dateString}`;
```
Comment (medikoo) [inline lib/plugins/aws/rollback.js:80]: I think here what should be overriden is just `serverless.getProvider('aws').deploymentTimestamp` and having that getters should start resolve the right deployment path (?)

File: test/unit/lib/plugins/aws/provider.test.js [hunk 5]
```diff
@@ -326,6 +327,28 @@ describe('AwsProvider', () => {
       return expect(awsProvider.options['aws-s3-accelerate']).to.be.undefined;
     });
   });
+
+  describe('#s3CodeArtifactsDirectoryPath', () => {
+    beforeEach(() => {
+      awsProvider.hooks.initialize();
+    });
+    it('should provide proper code artifacts S3 directory path', () => {
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/provider.test.js:335]: Any new tests should be based on [`runServerless`](https://github.com/serverless/test/blob/master/docs/run-serverless.md) util, e.g. see how `provider.getRegion()` is tested: https://github.com/serverless/serverless/blob/9e308bdf041c73eb2d27408ae4e03bc10fe9bf8a/test/unit/lib/plugins/aws/provider.test.js#L602-L637

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @remi00 and sorry that it took us so long to review that (for some reason it got lost in a process, and it's great you've reminded us)

In general it looks very good, I have just few suggestions

**Comment by remi00** [comment]
@wwwehr you mentioned in #9876 there are some changes already in the baseline that handle this. Could you point me to them? Looks like it was not obvious to me yesterday evening. @medikoo maybe you know and can help?

**Comment by remi00** [comment]
This is already far away from possibility of merging to main

**Pull request closed without merging**